File: soln-binary-search-problems/peak-element-index.java
/*
https://practice.geeksforgeeks.org/problems/peak-element/1
https://leetcode.com/problems/find-peak-element/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    public int findPeakElement(int[] nums) {
        return binarySearch(nums,0,nums.length-1);
    }
    
    public int binarySearch(int[] nums, int start, int end){
        while (start < end) {
            int mid = start+(end-start) / 2;
            if (nums[mid] > nums[mid + 1])
                end = mid;
            else
                start = mid + 1;
        }
        return start;
    }
}

File: C++/soln-binary-search-problems/peak-element-index.cpp
/**
 @author Farheen Bano

 Date 19-07-2021
  
 References-
 https://practice.geeksforgeeks.org/problems/peak-element/1
 https://leetcode.com/problems/find-peak-element/
*/

class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int left=0, right=nums.size()-1;
       while (left < right) {
            int mid = left+(right-left) / 2;
            if (nums[mid] > nums[mid + 1])
                right = mid;
            else
                left = mid + 1;
        }
        return left;
    }
};